src/seo.ts

```diff
@@ -1,20 +1,166 @@
+    if (scriptTag) {
+      const data = JSON.parse(scriptTag.textContent!);
+      if (this.chaptersData?.length && this.config.baseSegmentsUrl) {
+        data.hasPart = this.getClips();
+      }
+      data.transcript = this.transcriptData;
+      scriptTag.textContent = JSON.stringify(data);
+    }
+  }
+
+  private getClips(): Clip[] {
+    return this.chaptersData!.map((chapter) => {
+      return {
+        '@type': 'Clip',
+        name: chapter.name,
+        startOffset: chapter.startTime,
+        endOffset: chapter.endTime,
+        url: `${this.config.baseSegmentsUrl}${chapter.startTime}`,
+        ...(chapter.description && { description: chapter.description })
+      };
+    });
+  }
+
+  private static sendSEOStructuredData(SEOStructuredData: WithContext<VideoObject>): void {
+    window.parent.postMessage({ type: 'SEOStructuredData', SEOStructuredData }, '*');
+  }
+
+  private hasStructuredDataRequiredProperties(): boolean {
+    const name = this.player.sources.metadata.name;
+    const thumbnailUrl = this.player.sources.poster;
+    const uploadDate = this.player.sources.metadata.createdAt;
+    return !!(name && thumbnailUrl && uploadDate);
+  }
+
+  private static isPlayerIframeEmbeded(): boolean {
+    return window.self !== window.top;
+  }
 
   public static isValid(): boolean {
     return true;
   }
 
-  protected loadMedia(): void {
-    this.todo();
+  private injectStructureData(jsonLdData: WithContext<VideoObject>): void {
+    const script = document.createElement('script');
+    script.id = SEO_SCRIPT_ID;
+    script.setAttribute('type', 'application/ld+json');
+    script.textContent = JSON.stringify(jsonLdData);
+    document.head.appendChild(script);
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private get cuePointManager(): any {
+    return this.player.getService('kalturaCuepoints');
+  }
+
+  private registerCuePointTypes(): void {
+    this.cuePointManager.registerTypes([this.cuePointManager.CuepointType.CHAPTER, this.cuePointManager.CuepointType.CAPTION]);
+  }
+
+  private onTimedMetadataAdded({ payload }: TimedMetadataEvent): void {
+    const { KalturaCuePointType, KalturaThumbCuePointSubType } = this.cuePointManager;
+    const chapterData: CuePoint[] = [];
+    const captionData: CuePoint[] = [];
+
+    payload.cues.forEach((cue: CuePoint) => {
+      const { metadata } = cue;
+      if (metadata?.cuePointType === KalturaCuePointType.THUMB && metadata?.subType === KalturaThumbCuePointSubType.CHAPTER) {
+        chapterData.push(cue);
+      }
+      if (metadata?.cuePointType === KalturaCuePointType.CAPTION) {
+        captionData.push(cue);
+      }
+    });
+    if (chapterData.length) {
+      this.chaptersData = Seo.extractRelevantChaptersData(chapterData);
+    }
+    if (captionData.length) {
+      this.transcriptData = Seo.generateTranscriptFromCuePoints(captionData);
+      this.resolveTimedDataReadyPromise();
```

Comment:
Correct, the structure data injection  is divided into two phases 
In the first phase, it is injected as soon as the first metadata is available (i.e. the data independent of load media and the second sever call) regardless of whether captions/chapters are available 
in the second phase - as soon as the chapters/captions are available, the structure data is updated.

that way, There will be no waiting in vain, in case captions are not available